build.assets/tooling/cmd/update-plist-version/main.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Command update-plist-version updates the version fields of a bundle plist.
+//
+// The version fields updated in an Info.plist file are the CFBundleVersion and
+// CFBundleShortVersionString fields. If the version is not valid as per the
+// Apple specification then the fields will be set to 1.0. A valid version is 3
+// positive integers separated by dots. A semver with a pre-release tag is not
+// valid.
+//
+// This is intended to be used on the tsh.app Info.plist files. Standard
+// releases have the required version number. Pre-releases do not.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"howett.net/plist"
+)
+
+func main() {
+	if len(os.Args) < 3 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <version> <plist-file>...\n", filepath.Base(os.Args[0]))
+		os.Exit(1)
+	}
+
+	// A version can only be three positive integers separated by periods.
+	versionRE := regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+$`)
+	version := os.Args[1]
+	if !versionRE.MatchString(version) {
+		version = "1.0"
+	}
+
+	for _, filename := range os.Args[2:] {
+		err := replaceFile(filename, func(in, out *os.File) error {
+			return updateVersion(version, in, out)
+		})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "could not update version of %s: %v\n", filename, err)
+			os.Exit(1)
+		}
+	}
+}
+
+func updateVersion(version string, in, out *os.File) error {
+	dict := map[string]any{}
+	d := plist.NewDecoder(in)
+	if err := d.Decode(&dict); err != nil {
+		return err
```

Comment:
`trace.Wrap` and/or add some context to the error.

Same throughout.

Reply:
I would prefer not to dump stack traces on the output. Fine for daemons logging to files, but for user-oriented commands, I think it is just not that useful and sometimes obscures what is going on.

The errors have enough context to be understandable. For instance, this is the error when the plist cannot be decoded (I changed `dict` to `ict`):

```
could not update version of /home/camh/p/tp/teleport/build.assets/macos/tshdev/tsh.app/Contents/Info.plist: plist: error parsing XML property list: encountered unknown element ict
```
another error
```
could not update version of /home/camh/p/tp/teleport/build.assets/macos/tshdev/tsh.app/Contents/Info.plist: open /home/camh/p/tp/teleport/build.assets/macos/tshdev/tsh.app/Contents/Info.plist2487345849: permission denied
```

another error

```
could not update version of /home/camh/p/tp/teleport/build.assets/macos/tshdev/tsh.app/Contents/Info.plist: CFBundleShortVersionString not in plist file
```

It's a shame we need to do the `cd build.assets/tooling` before running `go run` which necessitates absolute paths, making that error line longer than I like, but it was going to be a pretty long path anyway :shrug:

And I'm not sure that any more context in a 100-line program with 3 functions is really needed for debugging. I may be wrong and am happy to be convinced but I'm seeing `trace.Wrap` as a net negative right now.

Reply:
trace.Wrap is arguable, `fmt.Errorf("brief context: %w", err)` less so.

I agree it's a small program, so not asking for further changes. In terms of general practice I'd frown on most naked `return err`.